Merge Transforms branch to develop

cgmillette opened the pull request:
- Brings transforms utilities to dev branch
- Postures develop branch for layout optimizations

cgmillette committed:
Initial commit aos_soa block dim 16, 32, 64, 128

cgmillette committed:
Refactor vector ordering utilities into separate file. Optimize sub-dword vector re-ordering.

CongMa13 committed:
Implement aos_soa_16xk_b32, soa_aos_16xk_b32 of vw=4

- Implement aos_soa_16xk_b32, soa_aos_16xk_b32 of vw=4
- Fix bugs

CongMa13 committed:
Implement aos_soa_32xk_b32 and soa_aos_32xk_b32 of VW = 4

* Implement aos_soa_32xk_b32 and soa_aos_32xk_b32 of VW = 4
* use vector_generator instead of unroll
* Fix bugs

cgmillette committed:
Add unit tests for vector util

cgmillette committed:
Add unit test for pack util

cgmillette committed:
Merge pull request #320 from cgmillette/aos_soa_dev

Add unit test setups for vector_util and pack_util

cgmillette committed:
Refactor vector_util interface. Add unit tests

CongMa13 committed:
Add unit tests for vector util

cgmillette committed:
Fix pack util bugs

cgmillette committed:
Update vector util description with 4 element examples

cgmillette committed:
Merge pull request #324 from cgmillette/aos_soa_dev

Optimize vector utils

CongMa13 committed:
Implement aos_soa_64k_b32, soa_aos_64k_b32 of VW=4

- Implement aos_soa_64k_b32, soa_aos_64k_b32 of VW=4

CongMa13 committed:
Implement aos_soa_128k_b32, soa_aos_128k_b32 of VW=4

- Implement aos_soa_128k_b32, soa_aos_128k_b32 of VW=4

CongMa13 committed:
Add AosToSoa, SoaToAos structs of VW=4

Move code from aos_soa_***xk_b32, soa_aos_***xk_b32 to structs AosToSoa
and SoaToAos.

CongMa13 committed:
Add AosToSoa<256, 4>, SoaToAos<256, 4> structs of VW=4

- Add AosToSoa<256, 4>, SoaToAos<256, 4> structs of VW=4

CongMa13 committed:
Add unit tests for pack util

CongMa13 committed:
Use new algorithm for aos to soa and soa to aos

- Use unified algorithm for 64k, 128k and 256k aos to soa and soa to aos

CongMa13 committed:
Add unit test for transfroms

CongMa13 committed:
Return generated value from vector_generator

Return generated value from vector_generator
- DPP
- Swizzle
- Permute

Replaced some duplicated code with functions

cgmillette committed:
Merge pull request #330 from CongMa13/aos_soa_dev

Implement AOS->SOAand SOA->AOS of VW4

cgmillette committed:
Fixup aos-soa vw=8 BlockDim 128, 256

cgmillette committed:
Merge pull request #333 from cgmillette/aos_soa_dev

Fixup aos-soa vw=8 BlockDim 128, 256

CongMa13 committed:
Move f64 support of corss-lane to higher level

- DPP
- Swizzle
- Permute

CongMa13 committed:
Add unit test to transforms unpack functions

Commit:
support transforms VW=4 on navi3

CongMa13 committed:
Support Aossoa VW=8 on navi3

CongMa13 committed:
Rename struct matrix to Memory2DArray

CongMa13 committed:
Add comment about Arch macros in config.hpp

This is a know issue that Arch macros are only defined during
device compiler pass.

CongMa13 committed:
Fixed test data issue in transforms unit tests

CongMa13 committed:
permute cross-lane ops add cache for ctrl op

CongMa13 committed:
Add macro ROCWMMA_GENERATE_UNIT_GTEST_SUITE

cgmillette committed:
Adjust blend and permute to cache given opCtrl code when applying to vectors. Also support f64

Commit:
SoaToAos VW=2

CongMa13 committed:
Solve a rebase conflict

CongMa13 committed:
Declare default Aossoa::exec, Soaaos::exec

CongMa13 committed:
Use vector_generator to replace forEach

Commit:
SoaToAos VW=8 Implementation

CongMa13 committed:
Rename class Memory2DArray to RegisterBank

- delete dead code dppElement

cgmillette committed:
Merge branch 'aos_soa_dev' into cache-ctrlOp

cgmillette committed:
Merge pull request #346 from cgmillette/cache-ctrlOp

ctrlOp cache with f64 support

Commit:
Re-added VW=2 AoS and SoA initializations

Commit:
Cleanup and fixes

Commit:
Addressed comments

dlangbe committed:
Merge pull request #349 from dlangbe/aos_soa_dev

SoA to AoS VW = 2, 8

mkarunan committed:
Add AOS to SOA for VW = 2 (#350)

cgmillette committed:
Refactor ColInlineVW to follow prioritize VWSegs->BlockKSegs->BlockDimSegs. Update ColOrthoVW with consistent notations. Update inline docs

dlangbe committed:
Wave32 AosToSoa VW 2

dlangbe committed:
SoaToAos VW 2 and 4 Wave32 support

cgmillette committed:
Update transforms to remove need for re-pack. Remove redundant, unused code

cgmillette committed:
Update unit test layout references for AOS

Commit:
Refactored genData functions

dlangbe committed:
Merge pull request #354 from dlangbe/aos_soa_dev

AosSoa Wave32 Support

Commit:
Separated unpack utility tests from transforms test

Commit:
formatting

cgmillette committed:
Adjust to changes with ColInlineVW

cgmillette committed:
Update SoaToAos 64, 4

cgmillette committed:
Re-order the SoaToAos specializations to match the AosToSoa

dlangbe committed:
Merge pull request #356 from dlangbe/aos_soa_dev

Unpack utility tests

cgmillette committed:
Fixup compiler / validation on gfx11

cgmillette committed:
Merge pull request #355 from cgmillette/col_inline_adjust

Adjust ColInlineVW layout

dlangbe committed:
AosToSoa VW = 16 (#367)

* AosToSoa VW=16

* SoaToAos VW=16

---------

Co-authored-by: dlangbe <dlangbeh@amd.com>

mkarunan committed:
optimize unpack util function (#371)

CongMa13 committed:
Move transforms structs into namespace Transforms (#366)

* Initial refactor of layout objects and ortho properties. Update rocwmma api implementations for apply transpose/datalayout

* Enable new large block layouts in rocWMMA

* Move transforms structs into namespace Transforms

- Move transforms driver into namespace TransformsImpl::AosSoaImpl
- Move other transforms structs into namespace TransformsImpl::AosSoaImpl

* clang-format transforms_impl.hpp

---------

Co-authored-by: Chris Millette <chris.millette@amd.com>
Co-authored-by: Christopher Millette <63608002+cgmillette@users.noreply.github.com>

cgmillette committed:
Fix unit tests (#372)

* fixed io_shape unit test (#370)

---------

Co-authored-by: Cong Ma <142121551+CongMa13@users.noreply.github.com>

cgmillette committed:
Merge branch 'develop' into aos_soa_dev

cgmillette committed:
remove using std::decay_t

cgmillette committed:
Fix link to device compiler pass

CongMa13 commented on library/include/rocwmma/internal/vector_util_impl.hpp:
```diff
@@ -406,22 +410,42 @@ namespace rocwmma
-                           ? Blend::UnpackWordHi::exec(get<Index>(v0), get<Index>(v1))
-                           : Blend::UnpackByteHi::exec(get<Index>(v0), get<Index>(v1));
-            };
+            if constexpr(ElementSize < 2u && PackedVecSize == 0)
```
Do we need a comment explains why it needs to be treated differently.

CongMa13 commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -155,14 +173,16 @@ namespace rocwmma
         static_assert(VecSize % 2 == 0, "VecSize must be a multiple of 2");
         using PackUtil = PackUtil<DataT>;
 
+        // TODO replace with dpp::move
```
mark.  Should we remove this TODO?

cgmillette replied:
Let's keep them, because they are still to-do. Will do a pass following the merge to clean these up.

CongMa13 commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -176,694 +196,2007 @@ namespace rocwmma
         auto hi = PackUtil::paddedPack(extractOdd(v));
 
         // TODO: label as rotateR64 for consistency?
+        // TODO replace with dpp::move
```
mark.  Should we remove this TODO?

CongMa13 commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -176,694 +196,2007 @@ namespace rocwmma
+                    // Step 2 : Unpack groups of 8
+                    result = unpackLoHi8(result);
+
+                    if constexpr (PackUtil::Traits::PackRatio >= VW)
```
only applied for vw=4,8. Does 16 needs it?

CongMa13 commented on library/include/rocwmma/internal/blend_impl.hpp:
```diff
@@ -102,18 +121,29 @@ namespace rocwmma
-                        = (reinterpret_cast<uint32_t&>(src1) & mask)
-                          | (reinterpret_cast<uint32_t&>(src0) & ~mask);
-                    return src0;
+                    // TODO:: WHY?
+                    // reinterpret_cast<uint32_t&>(src0)
+                    //     = (reinterpret_cast<uint32_t&>(src1) & maskCtrl)
+                    //       | (reinterpret_cast<uint32_t&>(src0) & ~maskCtrl);
+                    // return src0;
```
should we remove it?

CongMa13 commented on test/reference_impl.hpp:
```diff
@@ -322,54 +326,64 @@ namespace rocwmma
+                        uint32_t ele1 = ((uint32_t*)&blendSrc1)[index];
+                        // 0 <= Select < 4  element 0
+                        // 4 <= Select < 8  element 1
+                        // TODO uint32_t => DataT
```
Mark. Added by me. Should remove.

CongMa13 replied:
fixed in [6316756](https://github.com/ROCm/rocWMMA/pull/368/commits/631675695ab1709e09395b4ce3ccf2341aca2e9e)

CongMa13 commented on test/unit/vector_util_test/test/vector_util.cpp:
```diff
@@ -39,7 +39,7 @@ namespace rocwmma
 
         // Types: Base IOC + double
-        using Types = typename Base::TestTypes16;
+        using Types = std::tuple<int8_t, float16_t, bfloat16_t, float32_t, float64_t>;
```
use `typename Base::TestAllSizeTypes;`

CongMa13 replied:
fixed in [6316756](https://github.com/ROCm/rocWMMA/pull/368/commits/631675695ab1709e09395b4ce3ccf2341aca2e9e)

CongMa13 commented:
There were some building errors. Created a new commit to fix them.

CongMa13 committed:
fixed merge conflits

cgmillette committed:
Fix CP_BLK tests (#373)

CongMa13 committed:
Break the circular include

- Break the circular include
- Solved the warning "use of function template name with no prior
declaration in function call with explicit template arguments is a C++20
extension"

cgmillette committed:
Merge branch 'develop' into aos_soa_dev

CongMa13 reviewed (approved):
lgtm
